File: comparison/interface/interfaces.exs
#!/usr/bin/env elixir

# #####################################
# 영어 순서: a Subject -> Verbs -> Objects
# Elixir 설계 순서: a Subject -> Verbs -> Objects
# Elixir 구현 순서: Objects -> Verbs -> a Subject
# #####################################

# ###################################
# Objects
# ###################################
defmodule Pet do
  @callback eat() :: :ok
  @callback make_sound() :: :ok
end

defmodule Dog do
  @behaviour Pet

  def eat do
    IO.puts("The dog is eating.")
  end

  def make_sound do
    IO.puts("Woof! Woof!")
  end
end

defmodule Cat do
  @behaviour Pet

  def eat do
    IO.puts("The cat is eating.")
  end

  def make_sound do
    IO.puts("Meow! Meow!")
  end
end

defmodule Main do
  # ###################################
  # Verbs
  # ###################################
  def feed_pet(pet) when is_atom(pet) and pet in [Dog, Cat] do
    pet.eat()
  end

  def walk_the_pet(pet) when is_atom(pet) and pet in [Dog, Cat] do
    pet.make_sound()
  end

  # ###################################
  # a Subject
  # ###################################
  def main do
    dog = Dog
    cat = Cat

    feed_pet(dog)
    feed_pet(cat)
    walk_the_pet(dog)
    walk_the_pet(cat)
  end
end

# Apply
Main.main()

File: comparison/interface/interfaces.rs
#!/usr/bin/env rust-script

//! 영어 순서: a Subject -> Verbs -> Objects
//! Rust 설계 순서: a Subject -> Verbs -> Objects
//! Rust 구현 순서: Objects -> Verbs -> a Subject

// ////////////////////////////////////
// a Subject
// ////////////////////////////////////
fn main() {
    let dog = Dog;
    let cat = Cat;

    feed_pet1(&dog);
    feed_pet1(&cat);

    feed_pet2(&dog);
    feed_pet2(&cat);

    walk_the_pet1(&dog);
    walk_the_pet1(&cat);

    walk_the_pet2(&dog);
    walk_the_pet2(&cat);
}

// ////////////////////////////////////
// Verbs
// ////////////////////////////////////

// 라이브러리 개발 -> 제네릭 사용
fn feed_pet1<T: Pet>(pet: &T) {
    pet.eat();
}

// 응용 프로그램 코드 -> impl Trait 문법 사용
fn feed_pet2(pet: &impl Pet) {
    pet.eat();
}

// 라이브러리 개발 -> 제네릭 사용
fn walk_the_pet1<T: Pet>(pet: &T) {
    pet.make_sound();
}

// 응용 프로그램 코드 -> impl Trait 문법 사용
fn walk_the_pet2(pet: &impl Pet) {
    pet.make_sound();
}

// ////////////////////////////////////
// Objects
// ////////////////////////////////////
struct Dog;
struct Cat;

trait Pet {
    fn eat(&self);
    fn make_sound(&self);
}

impl Pet for Dog {
    fn eat(&self) {
        println!("The dog is eating.");
    }

    fn make_sound(&self) {
        println!("Woof! Woof!");
    }
}

impl Pet for Cat {
    fn eat(&self) {
        println!("The cat is eating.");
    }

    fn make_sound(&self) {
        println!("Meow! Meow!");
    }
}
